src/cluster/agglomerative.rs

```diff
@@ -0,0 +1,392 @@
+//! > In each case, if necessary some pair of subsets is merged to maintain the desired number of subsets.
+//! > Amortized time per insertion is O(Q log n); amortized time per deletion is O(Q log2 n). Space is linear.
+//!
+//! > FastPair. We further simplify conga lines by making separate singleton subsets for the objects moved to
+//! > new subsets by a deletion. This can alternately be viewed as a modification to the neighbor heuristic, in
+//! > which the initial construction of all nearest neighbors is replaced by a conga line computation, and in
+//! > which each insertion does not update previously computed neighbors. Its time both theoretically and in practice
+//! > is qualitatively similar to the neighbor heuristic, but it typically runs 30% or so faster.
+//!
+//! ## Interface
+//! Reference:
+//!     >>> from sklearn.cluster import AgglomerativeClustering
+//!     >>> import numpy as np
+//!     >>> X = np.array([[1, 2], [1, 4], [1, 0],
+//!     ...               [4, 2], [4, 4], [4, 0]])
+//!     >>> clustering = AgglomerativeClustering().fit(X)
+//!     >>> clustering
+//!     AgglomerativeClustering()
+//!     >>> clustering.labels_
+//!     array([1, 1, 1, 0, 0, 0])
+//!
+//! ## Example:
+//!
+//! ```
+//! use smartcore::linalg::naive::dense_matrix::*;
+//! use smartcore::math::num::RealNumber;
+//! use smartcore::cluster::agglomerative::*;
+//! use smartcore::algorithm::neighbour::fastpair::FastPair;
+//!
+//! // Iris data
+//! let x = DenseMatrix::from_2d_array(&[
+//!            &[5.1, 3.5, 1.4, 0.2],
+//!            &[4.9, 3.0, 1.4, 0.2],
+//!            &[4.7, 3.2, 1.3, 0.2],
+//!            &[4.6, 3.1, 1.5, 0.2],
+//!            &[5.0, 3.6, 1.4, 0.2],
+//!            &[5.4, 3.9, 1.7, 0.4],
+//!            &[4.6, 3.4, 1.4, 0.3],
+//!            &[5.0, 3.4, 1.5, 0.2],
+//!            &[4.4, 2.9, 1.4, 0.2],
+//!            &[4.9, 3.1, 1.5, 0.1],
+//!            &[7.0, 3.2, 4.7, 1.4],
+//!            &[6.4, 3.2, 4.5, 1.5],
+//!            &[6.9, 3.1, 4.9, 1.5],
+//!            &[5.5, 2.3, 4.0, 1.3],
+//!            &[6.5, 2.8, 4.6, 1.5],
+//!            &[5.7, 2.8, 4.5, 1.3],
+//!            &[6.3, 3.3, 4.7, 1.6],
+//!            &[4.9, 2.4, 3.3, 1.0],
+//!            &[6.6, 2.9, 4.6, 1.3],
+//!            &[5.2, 2.7, 3.9, 1.4],
+//!            ]);
+//!
+//! // Fit to data, with a threshold
+//! // example using FastPair, compute dendrogram
+//! let clustering = AggregativeFastPair::fit(&x).unwrap();
+//! let dendrogram = clustering.labels();
+//! ```
+//!
+use std::marker::PhantomData;
+
+use crate::algorithm::neighbour::fastpair::FastPair;
+use crate::error::Failed;
+use crate::linalg::Matrix;
+use crate::math::distance::euclidian::Euclidian;
+use crate::math::num::RealNumber;
+
+///
+/// Abstract trait for sequential, agglomerative, hierarchic, non-overlapping methods
+///
+pub trait SAHNClustering<T: RealNumber, M: Matrix<T>> {
```

Comment:
can you move method `fit` and `labels` directly into `AggregativeFastPair ` and remove this train for now? We do need a broader system of interface for all our algorithms but I think it would be best to have it in a separate PR.

Reply:
I can do this

Reply:
Just a note to clear: FastPair do not perform clustering, it is only a faster implementation to find the closest neighbour